tree/ntuple/v7/inc/ROOT/RColumn.hxx

```diff
@@ -81,26 +110,46 @@ public:
    void Connect(DescriptorId_t fieldId, RPageStorage *pageStorage);
 
    void Append(const RColumnElementBase &element) {
-      void *dst = fHeadPage.TryGrow(1);
-      if (dst == nullptr) {
-         Flush();
-         dst = fHeadPage.TryGrow(1);
-         R__ASSERT(dst != nullptr);
+      void *dst = fHeadPage[fHeadPageIdx].GrowUnchecked(1);
+
+      if (fHeadPage[fHeadPageIdx].GetNElements() == fApproxNElementsPerPage / 2) {
+         FlushShadowHeadPage();
       }
+
       element.WriteTo(dst, 1);
       fNElements++;
+
+      if (fHeadPage[fHeadPageIdx].GetNElements() == fApproxNElementsPerPage)
+         SwapHeadPages();
    }
 
    void AppendV(const RColumnElementBase &elemArray, std::size_t count) {
-      void *dst = fHeadPage.TryGrow(count);
-      if (dst == nullptr) {
+      // We might not have enough space in the current page. In this case, fall back to one by one filling.
+      if (fHeadPage[fHeadPageIdx].GetNElements() + count > fApproxNElementsPerPage) {
+         // TODO(jblomer): use (fewer) calls to AppendV to write the data page-by-page
          for (unsigned i = 0; i < count; ++i) {
             Append(RColumnElementBase(elemArray, i));
          }
          return;
       }
+
+      void *dst = fHeadPage[fHeadPageIdx].GrowUnchecked(count);
+
+      // The check for flushing the shadow page is more complicated than for the Append() case
+      // because we don't necessarily fill up to exactly fApproxNElementsPerPage / 2 elements;
+      // we might instead jump over the 50% fill level
+      if ((fHeadPage[fHeadPageIdx].GetNElements() <= fApproxNElementsPerPage / 2) &&
+          (fHeadPage[fHeadPageIdx].GetNElements() + count > fApproxNElementsPerPage / 2))
+      {
+         FlushShadowHeadPage();
+      }
+
       elemArray.WriteTo(dst, count);
       fNElements += count;
+
+      // Note that by the very first check, we cannot have filled more than fApproxNElementsPerPage elements
+      if (fHeadPage[fHeadPageIdx].GetNElements() == fApproxNElementsPerPage)
+         SwapHeadPages();
```

Comment:
Good point! Consolidated now in `SwapHeadPagesIfFull`.  For `FlushShadowWritePage`, it would be difficult to consolidate because the checks differ between `Append()` and `AppendV()`.